src/inputs/mod.rs

```diff
@@ -0,0 +1,451 @@
+    fn __setstate__(&mut self, state: &PyAny) -> PyResult<()> {
+        if let Ok(("KafkaInputConfig", brokers, group_id, topics, messages_per_epoch)) =
+            state.extract()
+        {
+            self.brokers = brokers;
+            self.group_id = group_id;
+            self.topics = topics;
+            self.messages_per_epoch = messages_per_epoch;
+            Ok(())
+        } else {
+            Err(PyValueError::new_err(format!(
+                "bad pickle contents for KafkaInputConfig: {state:?}"
+            )))
+        }
+    }
+}
+
+/// Use a user-defined function that returns an iterable as the input source.
+///
+///
+/// Args:
+///
+///     input_builder: An input_builder function that yields `AdvanceTo()` or `Emit()`
+///         with this worker's input. Must resume from the epoch specified.
+///
+/// Returns:
+///
+///     Config object. Pass this as the `input_config` argument to
+///     your execution entry point.
+#[pyclass(module = "bytewax.inputs", extends = InputConfig)]
+#[pyo3(text_signature = "(input_builder)")]
+pub(crate) struct ManualInputConfig {
+    input_builder: Py<PyAny>,
+}
+
+#[pymethods]
+impl ManualInputConfig {
+    #[new]
+    #[args(input_builder)]
+    fn new(input_builder: Py<PyAny>) -> (Self, InputConfig) {
+        (Self { input_builder }, InputConfig {})
+    }
+
+    /// Pickle as a tuple.
+    fn __getstate__(&self, py: Python) -> (&str, Py<PyAny>) {
+        ("ManualInputConfig", self.input_builder.clone_ref(py))
+    }
+
+    /// Egregious hack see [`SqliteRecoveryConfig::__getnewargs__`].
+    fn __getnewargs__(&self) -> (&str,) {
+        (&"",)
+    }
+
+    /// Unpickle from tuple of arguments.
+    fn __setstate__(&mut self, state: &PyAny) -> PyResult<()> {
+        if let Ok(("ManualInputConfig", input_builder)) = state.extract() {
+            self.input_builder = input_builder;
+            Ok(())
+        } else {
+            Err(PyValueError::new_err(format!(
+                "bad pickle contents for ManualInputConfig: {state:?}"
+            )))
+        }
+    }
+}
+
+/// Take a single data element or timestamp and feed it into dataflow
+/// Pump will be called in the worker's "main" loop to feed data in.
+pub trait Pump {
+    fn pump(&mut self);
+    fn input_time(&mut self) -> &u64;
+    fn input_remains(&mut self) -> bool;
+}
+
+/// Encapsulates the process of pulling data out of a Kafka
+/// stream and feeding it into Timely.
+struct KafkaPump {
+    consumer: BaseConsumer<CustomContext>,
+    rt: Runtime,
+    push_to_timely: InputHandle<u64, TdPyAny>,
+    desired_messages_per_epoch: u64,
+    current_messages_per_epoch: u64,
+    current_epoch: u64,
+    empty: bool,
+}
+
+impl KafkaPump {
+    fn new(config: PyRef<KafkaInputConfig>, push_to_timely: InputHandle<u64, TdPyAny>) -> Self {
+        let context = CustomContext;
+
+        let rt = tokio::runtime::Builder::new_current_thread()
+            .enable_all()
+            .build()
+            .unwrap();
+
```

Comment:
Link to [default values](https://github.com/edenhill/librdkafka/blob/master/CONFIGURATION.md)?